src/middleware/builtin.spec.ts

```diff
@@ -261,13 +261,180 @@ describe('directMention()', () => {
   });
 });
 
+describe('ignoreSelf()', () => {
+  it('should handle context missing error', async () => {
+    // Arrange
+    const fakeNext = sinon.fake();
+    const fakeBotUserId = undefined;
+    const fakeArgs = {
+      next: fakeNext,
+      context: { botUserId: fakeBotUserId },
+    } as unknown as MemberJoinedOrLeftChannelMiddlewareArgs;
+
+    const { ignoreSelf: getIgnoreSelfMiddleware, contextMissingPropertyError } = await importBuiltin();
+
+    // Act
+    const middleware = getIgnoreSelfMiddleware();
+    middleware(fakeArgs);
+    await delay();
+
+    // Assert
+    const expectedError = contextMissingPropertyError(
+      'botId',
+      'Cannot ignore events from the app without a bot ID. Ensure authorize callback returns a botId.',
+    );
+
+    const contextMissingError: ContextMissingPropertyError = fakeNext.getCall(0).lastArg;
+
+    assert.equal(contextMissingError.code, expectedError.code);
+    assert.equal(contextMissingError.missingProperty, expectedError.missingProperty);
+  });
+
+  it('should immediately call next(), because incoming middleware args don\'t contain event', async () => {
+    // Arrange
+    const fakeNext = sinon.fake();
+    const fakeBotUserId = 'BUSER1';
+    const fakeArgs = {
+      next: fakeNext,
+      context: { botUserId: fakeBotUserId },
+      command: {
+        command: '/fakeCommand',
+      },
+    } as unknown as CommandMiddlewareArgs;
+
+    const { ignoreSelf: getIgnoreSelfMiddleware } = await importBuiltin();
+
+    // Act
+    const middleware = getIgnoreSelfMiddleware();
+    middleware(fakeArgs);
+    await delay();
+
+    // Assert
+    assert(fakeNext.calledOnce);
+  });
+
+  it('should look for an event identified as a bot message from the same bot ID as this app and skip it', async () => {
+    // Arrange
+    const fakeNext = sinon.fake();
+    const fakeBotUserId = 'BUSER1';
+    // TODO: Fix typings here
+    const fakeArgs = {
+      next: fakeNext,
+      context: { botUserId: fakeBotUserId, botId: fakeBotUserId },
+      event: {
+        type: 'message',
+        subtype: 'bot_message',
+        bot_id: fakeBotUserId,
+      },
+      message: {
+        type: 'message',
+        subtype: 'bot_message',
+        bot_id: fakeBotUserId,
+      },
+    } as any;
```

Comment:
Unfortunately, I wasn't able to provide the correct type here :(
If someone could help, it would be great!